From ca6efb3a7122b8ae8388f2c78b834f0b3d3c40c5 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Fri, 8 Oct 2021 09:16:06 +0700
Subject: [PATCH] media: rzg2l-cru: correct the slot that has been transferred
 to memory bank

AMnMBS.MBSTS's value indicates the memory bank to be used for the next
frame. So recalculate slot value after getting from AMnMBS register to
receive correct value.
With this action, the latency reduces dramatically. It is about 250ms
reduced.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/media/platform/rzg2l-cru/rzg2l-dma.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/media/platform/rzg2l-cru/rzg2l-dma.c b/drivers/media/platform/rzg2l-cru/rzg2l-dma.c
index d27a1bed6954..341f710e7cda 100644
--- a/drivers/media/platform/rzg2l-cru/rzg2l-dma.c
+++ b/drivers/media/platform/rzg2l-cru/rzg2l-dma.c
@@ -1035,6 +1035,15 @@ static irqreturn_t rzg2l_cru_irq(int irq, void *data)
 	amnmbs = rzg2l_cru_read(cru, AMnMBS);
 	slot = amnmbs & AMnMBS_MBSTS;
 
+	/*
+	 * AMnMBS.MBSTS indicates the destination of Memory Bank (MB).
+	 * Recalculate to get the current transfer complete MB.
+	 */
+	if (slot == 0)
+		slot = HW_BUFFER_NUM - 1;
+	else
+		slot--;
+
 	/*
 	 * To hand buffers back in a known order to userspace start
 	 * to capture first from slot 0.
-- 
2.25.1

